typing/typedecl_unboxed.ml

```diff
@@ -51,3 +51,254 @@ let get_unboxed_type_representation env ty =
+        else if Path.same p Predef.path_lazy_t then
+          (* Lazy values can 'shortcut' the lazy block, and thus have many
+             different tags. When Config.flat_float_array, they
+             cannot be floats, so we might want to refine that if there
+             are strong use-cases. *)
+          any
+        else if Path.same p Predef.path_nativeint
+             || Path.same p Predef.path_int32
+             || Path.same p Predef.path_int64 then
+          block_shape [Obj.custom_tag]
+        else begin
+          let head_callstack = Callstack.head callstack_map ty in
+          if Callstack.visited p head_callstack then
+            let loc =
+              match Env.find_type p env with
+              | decl -> Some decl.type_loc
+              | exception Not_found -> None in
+            Location.raise_errorf ?loc
+              "Cyclic type expansion during [@unboxed] verification.@ \
+               %a appears unboxed at the head of its own definition."
+              Path.print p
+          else match Env.find_type_descrs p env, Env.find_type p env with
+          | exception Not_found -> any
+          | descr, decl ->
+              of_typedescr env descr decl ~args callstack_map
+                (Callstack.visit p head_callstack)
+        end
+    | Ttuple _ -> block_shape [0]
+    | Tarrow _ | Tpackage _ | Tobject _ | Tnil | Tvariant _ -> (* XXX *) any
+    | Tlink _ | Tsubst _ | Tpoly _ | Tfield _ ->
+        assert false
+
+
+  and of_typedescr env ty_descr ty_decl ~args
+                   callstack_map new_callstack =
+    (* we called scrape_ty on the type expression in the function above, thus
+       it isnt a single constructor unboxed type.
+       TODO check what happens in the [@@unboxed] case (on the whole definition)
+     *)
+    let of_type_with_params ty =
+      (* We instantiate the formal type variables with the
+         type expression parameters at use site.
+
+         We build a callstack for this new type by mapping all new
+         nodes, corresponding to the spine of 'ty', to the current
+         call stack. *)
+      let ty = ty_of_poly ty in
+      let params = ty_decl.type_params in
+      let ty = Ctype.apply env params ty args in
+      let callstack_map =
+        Callstack.extend callstack_map ty new_callstack in
+      of_type env ty callstack_map
+    in
+    match ty_descr with
+    | Type_abstract ->
+       begin match ty_decl.type_manifest with
+       | None -> any
+       | Some ty -> of_type_with_params ty
+       end
+    | Type_record (_, Record_regular) -> block_shape [0]
+    | Type_record (_, Record_float) -> block_shape [Obj.double_array_tag]
+    | Type_record ([{lbl_arg = ty; _}], Record_unboxed _) ->
+        of_type_with_params ty
+    | Type_record (_, Record_unboxed _)
+    | Type_record (_, Record_inlined _)
+    | Type_record (_, Record_extension _) -> assert false
+    | Type_open -> block_shape [0]
+    | Type_variant ([],_) -> none
```

Comment:
Maybe renaming both `any` to `any_shape` and `none` to `empty_shape` would read better.

Reply:
In general they will be used as `Head_shape.{any,none,empty}`, so it will read well. We happen to be inside the module so it's a bit more demanding on the naming.

Reply:
I think those values won't be in the signature of `Head_shape` so only used in the implementation.